libkbfs/dirty_file.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2016 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"fmt"
+	"sync"
+)
+
+// syncBlockState represents that state of a block with respect to
+// whether it's dirty or currently being synced.  There can be six states:
+//  1) Not dirty or being synced
+//  2) Dirty and awaiting a sync
+//  3) Being synced and not yet re-dirtied: any write needs to
+//     make a copy of the block before dirtying it.  Also, all writes must be
+//     deferred.
+//  4) Being synced and already re-dirtied: no copies are needed, but all
+//     writes must still be deferred.
+//  5) Finished syncing, and not yet re-dirtied, but the rest of the
+//     sync hasn't finished yet.
+//  5) Finished syncing, and already re-dirtied, but the rest of the
+//     sync hasn't finished yet.
+type syncBlockState int
+
+const (
+	blockNotDirty syncBlockState = iota
+	blockDirty
+	blockSyncingNotDirty
+	blockSyncingAndDirty
+	blockSyncedAndDirty
+	blockSyncedNotDirty
+)
+
+// dirtyFile represents a particular file that's been written to, but
+// has not yet completed syncing its dirty blocks to the server.
+type dirtyFile struct {
+	lock sync.Mutex
```

Comment:
Nothing in this struct blocks on I/O -- it is only used for state tracking and everything returns immediately.  In fact, all the uses of this structure are already covered under `folderBlockOps.blockLock`, I believe -- I only added this as a fail safe in case the struct was ever used anywhere else.  I believe locking/unlocking will basically always be a no-op.  Does that make sense?